src/Fixer/ControlStructure/NoBreakCommentFixer.php

```diff
@@ -0,0 +1,309 @@
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'There must be a comment when fall-through is intentional in a non-empty case body.',
+            [
+                new CodeSample("<?php\nswitch (\$foo)\n{    case 1:\n        foo();\n    case 2:\n        bar();\n}"),
+                new CodeSample("<?php\nswitch (\$foo)\n{    case 1:\n        foo();\n        // no break\n        break;\n    case 2:\n        bar();\n}"),
+            ],
+            'Adds a "no break" comment before fall-through cases, and removes it if there is no fall-through.'
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAnyTokenKindsFound([T_CASE, T_DEFAULT]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $position => $token) {
+            if ($token->isGivenKind([T_CASE, T_DEFAULT])) {
+                $this->fixCase($tokens, $position);
+            }
+        }
+    }
+
+    /**
+     * @param Tokens $tokens
+     * @param int    $casePosition
+     */
+    private function fixCase(Tokens $tokens, $casePosition)
+    {
+        $empty = true;
+        $fallThrough = true;
+        $commentPosition = null;
+        for ($i = $tokens->getNextTokenOfKind($casePosition, [':', ';']) + 1, $max = count($tokens); $i < $max; ++$i) {
+            if ($tokens[$i]->isGivenKind([T_SWITCH, T_IF, T_ELSE, T_ELSEIF, T_FOR, T_FOREACH, T_WHILE, T_DO, T_FUNCTION, T_CLASS])) {
+                $empty = false;
+                $i = $this->getStructureEnd($tokens, $i);
+
+                continue;
+            }
+
+            if ($tokens[$i]->isGivenKind([T_BREAK, T_CONTINUE, T_RETURN])) {
+                $fallThrough = false;
+
+                continue;
+            }
+
+            if ($tokens[$i]->equals('}') || $tokens[$i]->isGivenKind(T_ENDSWITCH)) {
+                if (null !== $commentPosition) {
+                    $this->removeComment($tokens, $commentPosition);
+                }
+
+                break;
+            }
+
+            if ($this->isNoBreakComment($tokens[$i])) {
+                $commentPosition = $i;
+
+                continue;
+            }
+
+            if ($tokens[$i]->isGivenKind([T_CASE, T_DEFAULT])) {
+                if (!$empty && $fallThrough) {
+                    if (null !== $commentPosition && $tokens->getPrevNonWhitespace($i) !== $commentPosition) {
+                        $this->removeComment($tokens, $commentPosition);
+                        $commentPosition = null;
+                    }
+
+                    if (null === $commentPosition) {
+                        $this->insertCommentAt($tokens, $i);
+                    } else {
+                        $this->ensureNewLineAt($tokens, $commentPosition);
```

Comment:
alright, if you want to keep/add/detect the indent than please add priority tests for at least `braces` and `indentation_type`